### PR TITLE
Implement a simple stack dump on panic.

### PR DESCRIPTION
**Pull request by stillson**
This does no symbol resolution, just walks up the stack frame and prints the return addresses, to a maximum of 10 calls.

### CODE REVIEW

File: src/backtrace.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+pub(crate) fn backtrace() -> [Option<usize>; MAX_BACKTRACE_ADDRESSES] {
+    // Safe because we are just reading a register
+    let fp = unsafe {
+        let mut _tmp: u64;
```
Comment (atulkharerivos) [inline src/backtrace.rs:20]: Nit: _ should be used only for unused variables (unless it's being used to pacify the compiler in this context).

File: src/backtrace.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    let mut skip = 2;
+
+    // Safe because we trust the linker placed these symbols correctly.
+    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+    let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+
+    let mut fp = fp;
+    let mut old_address = 0;
+    loop {
+        // safe because we are only reading in the valid stack trace area
```
Comment (atulkharerivos) [inline src/backtrace.rs:40]: Nit: s/safe/Safe/

**Review by atulkharerivos — COMMENTED** [review]
Nice...is this safe to use even if `-Cforce-frame-pointers` is accidentally omitted?

File: src/backtrace.rs (hunk 2 above)
Comment (stillson) [reply src/backtrace.rs]: fixed

File: src/backtrace.rs (hunk 1 above)
Comment (stillson) [reply src/backtrace.rs]: fixed

**Comment by stillson** [comment]
I believe it's ok if you don't force frame-pointers. The only dangerous thing it does is pull memory off the stack, and it check if it is in stack bounds first.

File: src/backtrace.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    // Safe because we trust the linker placed these symbols correctly.
+    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+    let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+
+    let mut fp = fp;
+    loop {
+        // Safe because we are only reading in the valid stack trace area
+        // Also, this is only called on panic
+
+        if fp < stack_start || fp > stack_end {
```
Comment (rbradford) [inline src/backtrace.rs:42]: Would it be interesting to know when the fp points outside the expected range (stack corruption?) could there be a `println!()` of a message here (or maybe return a `Result<...>` and handle it elsewhere.)

File: src/backtrace.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        tmp
+    };
+
+    backtrace_internal(fp)
+}
+
+fn backtrace_internal(fp: u64) -> [Option<usize>; MAX_BACKTRACE_ADDRESSES] {
+    let mut result = [None; MAX_BACKTRACE_ADDRESSES];
+    let mut index = 0;
+    let mut skip = 2;
```
Comment (rbradford) [inline src/backtrace.rs:31]: Naive question: is this here to skip including this code itself?

File: src/backtrace.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+        if fp < stack_start || fp > stack_end {
+            break;
+        }
+
+        let address = unsafe { (fp as *const u64).offset(-1).read_volatile() }; // RA/PC
+        fp = unsafe { (fp as *const u64).offset(-2).read_volatile() }; // next FP
+
+        if address == 0 {
+            break;
```
Comment (rbradford) [inline src/backtrace.rs:50]: Again, is this something that we could usefully report an error out on?

File: attestation/BUILD [hunk 6]
```diff
@@ -27,6 +27,9 @@ rust_library(
         "@salus-index//:hex",
         "@salus-index//:typenum",
     ],
+    rustc_flags = [
+        "-Cforce-frame-pointers",
+    ],
```
Comment (abrestic-rivos) [inline attestation/BUILD:32]: instead of putting this in every single `BUILD` file, could you put it in the `.bazelrc` or similar?

**Review by abrestic-rivos — COMMENTED** [review]
i'm not convinced we should be rolling our own backtrace here... have you tried getting something like `mini-backtrace` to work?

**Comment by rbradford** [comment]
> mini-backtrace

I also found https://docs.rs/unwinding/latest/unwinding/ which doesn't pull in the libunwind (i.e. it's pure Rust.)

File: src/backtrace.rs (hunk 4 above)
Comment (stillson) [reply src/backtrace.rs:31]: That, and the panic code that is always the same.

File: attestation/BUILD (hunk 6 above)
Comment (stillson) [reply attestation/BUILD:32]: Well examine

File: src/backtrace.rs (hunk 3 above)
Comment (stillson) [reply src/backtrace.rs:42]: this might be best handled with a enum for all the error types

File: src/backtrace.rs (hunk 5 above)
Comment (stillson) [reply src/backtrace.rs:50]: again, enum

File: src/backtrace.rs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    let fp = unsafe {
+        let mut tmp: u64;
+        asm!("mv {0}, x8", out(reg) tmp);
+        tmp
+    };
+
+    backtrace_internal(fp)
+}
+
+fn backtrace_internal(fp: u64) -> [BTReturn; MAX_BACKTRACE_ADDRESSES] {
```
Comment (dgreid) [inline src/backtrace.rs:36]: 'fp' is dereferenced as a raw pointer, so this function shouldn't be safe.

Comment (stillson) [reply src/backtrace.rs]: fixed.

File: src/backtrace.rs (hunk 3 above)
Comment (stillson) [reply src/backtrace.rs:50]: fixed

File: src/backtrace.rs (hunk 5 above)
Comment (stillson) [reply src/backtrace.rs:60]: fixed

File: src/backtrace.rs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+pub(crate) fn backtrace() -> [BTReturn; MAX_BACKTRACE_ADDRESSES] {
+    // Safe because we are just reading a register
+    let fp = unsafe {
+        let mut tmp: u64;
+        asm!("mv {0}, x8", out(reg) tmp);
+        tmp
+    };
+
+    unsafe { backtrace_internal(fp) }
```
Comment (atulkharerivos) [inline src/backtrace.rs:33]: Needs a comment explaining why it's safe, and also corresponding documentation on the caller requirements in `backtrace_internal()`. 

Probably something like:

```
// # Safety
// The caller must ensure that fp is a valid address on the stack frame.
unsafe fn backtrace_internal(fp: u64) -> [BTReturn; MAX_BACKTRACE_ADDRESSES] {
```

File: src/backtrace.rs [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    let mut index = 0;
+    let mut skip = 2;
+
+    // Safe because we trust the linker placed these symbols correctly.
+    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+    let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+
+    let mut fp = fp;
+    loop {
+        // Safe because we are only reading in the valid stack trace area
```
Comment (atulkharerivos) [inline src/backtrace.rs:47]: Nit: Comment should be moved a few lines below.

File: src/backtrace.rs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub(crate) enum BTReturn {
+    Some(usize),
+    None,
+    BadFp(usize),
+    BadAddress(usize),
```
Comment (atulkharerivos) [inline src/backtrace.rs:22]: Nit: Since both u64 and usize are are 64-bit, using u64 will avoid the requirement for casting.

File: attestation/BUILD (hunk 6 above)
Comment (stillson) [reply attestation/BUILD:32]: Fixed. added to .bazelrc

File: src/backtrace.rs (hunk 8 above)
Comment (stillson) [reply src/backtrace.rs:33]: fixed

File: src/backtrace.rs (hunk 9 above)
Comment (stillson) [reply src/backtrace.rs:47]: fixed

File: src/backtrace.rs (hunk 10 above)
Comment (stillson) [reply src/backtrace.rs:22]: fixed

File: src/backtrace.rs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+pub const MAX_BACKTRACE_ADDRESSES: usize = 10;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub(crate) enum BTReturn {
+    Some(u64),
```
[the remark this answers is not in the document]
Comment (stillson) [reply src/backtrace.rs:19]: Makes sense. fixed

File: src/main.rs [hunk 12]
```diff
@@ -77,6 +79,31 @@ use umode::UmodeTask;
...
+        let bt = backtrace();
+        println!(
+            "panic occurred in file '{}' at line {}",
+            location.file(),
+            location.line()
+        );
+        println!("panic backtrace:");
+        for frame in bt {
+            match frame {
+                BTReturn::Some(x) => {
```
[the remark this answers is not in the document]
Comment (stillson) [reply src/main.rs:92]: yes, good idea. fixed.

File: src/backtrace.rs [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+pub const MAX_BACKTRACE_ADDRESSES: usize = 10;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub(crate) enum BTReturn {
```
[the remark this answers is not in the document]
Comment (stillson) [reply src/backtrace.rs:18]: Fixed

File: src/backtrace.rs [hunk 14]
```diff
@@ -0,0 +1,95 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use alloc::fmt::{Display, Formatter, Result};
+/// Get an array of backtrace addresses.
```
Comment (rbradford) [inline src/backtrace.rs:6]: nit: Not sure if the project enforces a particular style for use statements (grouping by external/internal) but I doesn't look right to have a comment in the middle here.

Comment (stillson) [reply src/backtrace.rs:6]: fixed

File: src/backtrace.rs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    let mut fp = fp;
+    loop {
+        if fp < stack_start || fp > stack_end {
+            result[index] = BTReturnAddress::InvalidFramePointer(fp);
+            break;
+        }
+
+        // Safe because we are only reading in the valid stack trace area
+        let address = unsafe { (fp as *const u64).offset(-1).read_volatile() }; // RA/PC
+        fp = unsafe { (fp as *const u64).offset(-2).read_volatile() }; // next FP
```
Comment (atulkharerivos) [inline src/backtrace.rs:74]: Sorry if this a duplicate, but my Github comment rendering has been really weird.

My test could be flawed, but what if fp == stack_start when we begin execution? Basically, we are effectively addressing *(fp - 16) on line 74, meaning that we can underflow the bounds of the stack?

```
fn assert_stack() {
     // Safe because we trust the linker placed these symbols correctly.
     let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
     let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
     let fp = stack_start;
     //  Same as if fp < stack_start || fp > stack_end on 67
     if fp >= stack_start && fp <= stack_end {
        let address = unsafe { (fp as *const u64).offset(-1) as u64 };
        let address = unsafe { (fp as *const u64).offset(-2) as u64 };
        assert!(address < stack_end);
        assert!(address > stack_start);
     }
}
```

Comment (atulkharerivos) [reply src/backtrace.rs:74]: On a somewhat related note, can we treat the stack as an as `&u64` slice that begins with stack_start, and with `len == (stack_end - stack_start) / 8`? That way, we can do the unsafe{} cast once, and will get bounds check for free, instead of relying on multiple unsafe checks.

Comment (atulkharerivos) [reply src/backtrace.rs:74]: Actually, there's no math involved, except for fp - stack_start. You can probably even mark the method as safe (but that's a separate discussion). 

Also, you simplify the code by replacing the loop with a `while index < MAX_BACKTRACE_ADDRESSES`.

For example, the following appears to be the equivalent of your code:

```
// Safety
// The caller must ensure that fp is a valid address on the stack frame.
unsafe fn backtrace_internal(fp: u64) -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
    use s_mode_utils::print::CONSOLE;
    use s_mode_utils::print::*;
    let mut result = [BTReturnAddress::End; MAX_BACKTRACE_ADDRESSES];
    let mut index = 0;
    let mut skip = 2;

    // Safe because we trust the linker placed these symbols correctly.
    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
    let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
    let mut fp  = fp;
    let stack = unsafe {core::slice::from_raw_parts(stack_start as *const u64, (stack_end - stack_start) as usize)};
    loop {  
        if fp < (stack_start+2) || fp > stack_end {
            result[index] = BTReturnAddress::InvalidFramePointer(fp);
            break;
        }

        let offset = ((fp - stack_start - 2)/8) as usize;
        if offset > 0 && let Some(address) = stack.get(offset) && let Some(current_fp) = stack.get(offset-1) {
            fp = *current_fp;
            let address = *address;     
            if address == 0 {
                result[index] = BTReturnAddress::InvalidReturnAddress(fp);
                break;
            }
            
            if skip > 0 {
                skip -= 1;
                continue;
            }
    
            result[index] = BTReturnAddress::ReturnAddress(address);
            index += 1;
    
            if index >= MAX_BACKTRACE_ADDRESSES {
                break;
            }        
        } else {
            result[index] = BTReturnAddress::InvalidReturnAddress(fp);
            break;
        }
    }

    result
}

```

**Comment by rbradford** [comment]
I think this needs revisiting with per-vCPU VA mapped stack work that is in #290

File: src/main.rs [hunk 16]
```diff
@@ -78,6 +80,21 @@ use umode::UmodeTask;
...
 fn panic(info: &core::panic::PanicInfo) -> ! {
+    if let Some(location) = info.location() {
+        println!(
+            "panic occurred in file '{}' at line {}",
+            location.file(),
+            location.line()
+        );
+    } else {
+        println!("panic occurred but can't get location information...");
+    }
```
Comment (glg-rv) [inline src/main.rs:91]: This information is already printed on the "info" down later.

**Comment by glg-rv** [comment]
> I think this needs revisiting with per-vCPU VA mapped stack work that is in #290

Yes AFAICT though the main thing to change is where to get the stack end/start. the rest should be hopefully fine.

File: src/backtrace.rs [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+// Safety
+// The caller must ensure that fp is a valid address on the stack frame.
+unsafe fn backtrace_internal(fp: u64) -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
+    let mut result = [BTReturnAddress::End; MAX_BACKTRACE_ADDRESSES];
+    let mut index = 0;
+    let mut skip = 2;
+
+    // Safe because we trust the linker placed these symbols correctly.
+    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
```
Comment (atulkharerivos) [inline src/backtrace.rs:62]: I still think that using a &[u64] slice is much better approach than trying to shift bits and check for boundary conditions. Using a slice only requires _stack_end and _stack_start to be right (and it's a given implicit assumption in both cases). However, when using slices, reviewers (and in the future maintainers) don't have to worry about getting anything else right  in the unsafe blocks (since there are none), and all checks are enforced by the compiler.

Comment (rbradford) [reply src/backtrace.rs:62]: And with the slice approach you can even iterate over it rather than the current `loop {}` with the `offset` calculation which is more of an index

Comment (atulkharerivos) [reply src/backtrace.rs:62]: Furthermore, it will make it possible to hoist the skip logic by simply returning the slice, and then using slice.iter().skip(2) at the caller site, which simplifies the code even further.

File: src/backtrace.rs (hunk 15 above)
Comment (stillson) [reply src/backtrace.rs:74]: fixed

File: src/main.rs (hunk 16 above)
Comment (stillson) [reply src/main.rs:91]: fixed

File: src/backtrace.rs (hunk 17 above)
Comment (atulkharerivos) [reply src/backtrace.rs:62]: No, we don't need to create a stack iterator because you get it for free after creating a slice. For example, I think you can simply return `[BTReturnAddress; MAX_BACKTRACE_ADDRESSES]` and then use iter() in the return value.

For example:

`
let stack_trace = unsafe {backtrace_internal(fp)};
stack_trace.to_iter().skip(2).for_each(|entry_type|) {
   // Handle entry type.
}
`

Also, notice how we use unsafe{} exactly once below with the linker defined symbols, and then rely on the compiler to enforce the rest without having to worry about whether we got it right.

/```
/ Safety
// The caller must ensure that fp is a valid address on the stack frame.
unsafe fn backtrace_internal(fp: u64) -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
    let mut result = [BTReturnAddress::End; MAX_BACKTRACE_ADDRESSES];
    let mut index = 0;
    let mut skip = 2;

    // Safe because we trust the linker placed these symbols correctly.
    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
    let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
    let mut fp  = fp;
    let stack = unsafe {core::slice::from_raw_parts(stack_start as *const u64, (stack_end - stack_start) as usize)};
    loop {  
        if fp < (stack_start+2) || fp > stack_end {
            break;
        }

        let offset = ((fp - stack_start - 2)/8) as usize;
        if offset > 0 && let Some(address) = stack.get(offset) && let Some(current_fp) = stack.get(offset-1) {
            fp = *current_fp;
            let address = *address;     
            if address == 0 {
                break;
            }
            
            result[index] = BTReturnAddress::ReturnAddress(address);
            index += 1;
    
            if index >= MAX_BACKTRACE_ADDRESSES {
                break;
            }        
        } else {
            break;
        }
    }

    result
}
```

File: src/backtrace.rs [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+            stack_start,
+            stack_end,
+            stack,
+        }
+    }
+}
+
+impl Iterator for BackTrace {
+    type Item = BTReturnAddress;
+    fn next(&mut self) -> Option<Self::Item> {
```
[the remark this answers is not in the document]
Comment (stillson) [reply src/backtrace.rs:69]: fixed

**Review by atulkharerivos — APPROVED** [review]

File: src/backtrace.rs [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+        }
+    }
+}
+
+//pub(crate) fn backtrace() -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
+pub(crate) fn backtrace() -> BackTrace {
+    // Safe because we are just reading a register
+    let fp = unsafe {
+        let mut tmp: u64;
+        asm!("mv {0}, x8", out(reg) tmp);
```
Comment (abrestic-rivos) [inline src/backtrace.rs:100]: s/x8/fp

File: src/backtrace.rs [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+            }
+            self.fp = None;
+            Some(BTReturnAddress::InvalidReturnAddress(fp))
+        } else {
+            None
+        }
+    }
+}
+
+//pub(crate) fn backtrace() -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
```
Comment (abrestic-rivos) [inline src/backtrace.rs:95]: stale comment?

File: src/backtrace.rs [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+            stack,
+        }
+    }
+}
+
+impl Iterator for BackTrace {
+    type Item = BTReturnAddress;
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(mut fp) = self.fp {
+            if fp < self.stack_start + 2 || fp > self.stack_end {
```
Comment (abrestic-rivos) [inline src/backtrace.rs:71]: why the `+2`?

File: src/backtrace.rs [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+                self.fp = None;
+                return Some(BTReturnAddress::InvalidFramePointer(fp));
+            }
+
+            let offset = ((fp - self.stack_start - 2) / 8) as usize;
```
Comment (abrestic-rivos) [inline src/backtrace.rs:76]: replace the 8 with `size_of::<u64>()`

File: src/backtrace.rs [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+            if offset > 0 && let Some(address) = self.stack.get(offset) && let Some(current_fp) = self.stack.get(offset-1) {
```
Comment (abrestic-rivos) [inline src/backtrace.rs:77]: should we enforce that the stack only go up? i.e. next fp must be > previous fp

Comment (rbradford) [inline src/backtrace.rs:77]: Perhaps this could be a method

```rust
fn next_address(&self, fp: u64) -> Option<(u64, u64> {
  let offset = ((fp - self.stack_start - 2) / 8) as usize;
  if offset == 0 {
     return None;
  }
  let address = self.stack.get(offset)?;
  let current_fp = self.stack.get(offset - 1)?;

  Some(address, current_fp)
```

That means you can then just do `if let Some(address, current_fp) = self.next_address()`

File: src/backtrace.rs (hunk 20 above)
Comment (stillson) [reply src/backtrace.rs:95]: yes, fixed

File: src/backtrace.rs (hunk 19 above)
Comment (stillson) [reply src/backtrace.rs:100]: fixed

File: src/backtrace.rs (hunk 21 above)
Comment (stillson) [reply src/backtrace.rs:71]: Not needed anymore. gone.

File: src/backtrace.rs (hunk 22 above)
Comment (stillson) [reply src/backtrace.rs:76]: fixed

File: src/backtrace.rs (hunk 23 above)
Comment (stillson) [reply src/backtrace.rs:77]: I can see some cases of asynchronous code where that might not be true. You might catch some strange bugs. Worth doing for now.

Comment (stillson) [reply src/backtrace.rs:77]: fixed

[the remark this answers is not in the document]
Comment (stillson) [reply src/backtrace.rs:77]: fixed

Comment (stillson) [reply src/backtrace.rs:77]: Actually, kernel_init uses an odd spot in the stack. leaving out for now, since this seems to cause an issue.

Comment (stillson) [reply src/backtrace.rs:77]: The problem with this is that you lose information (if offset or offset -1 are not in the stack).

Comment (rbradford) [reply src/backtrace.rs]: I'm not sure I follow? If either of `offset` or `offset - 1` some part of the if statement will be None and not executed. This function does the same thing by prematurely returning `None` via ?

File: src/backtrace.rs [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+impl Iterator for BackTrace {
+    type Item = BTReturnAddress;
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(mut fp) = self.fp {
+            if fp < self.stack_start || fp > self.stack_end {
+                self.fp = None;
+                return Some(BTReturnAddress::InvalidFramePointer(fp));
+            }
+
+            let offset = ((fp - self.stack_start - 2) / (size_of::<u64>() as u64)) as usize;
```
Comment (abrestic-rivos) [inline src/backtrace.rs:76]: why the `- 2`?

File: src/backtrace.rs [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+            if offset > 0 &&
```
Comment (abrestic-rivos) [inline src/backtrace.rs:77]: this is going to be true unless `fp == self.stack_start`. i think if you actually want to catch an `fp` below `stack_start` you need to do a `checked_sub()` above

File: src/main.rs [hunk 26]
```diff
@@ -79,6 +81,19 @@ use umode::UmodeTask;
...
+    if let Some(location) = info.location() {
+        println!(
+            "panic occurred in file '{}' at line {}",
+            location.file(),
+            location.line()
+        );
+    }
+
+    println!("panic backtrace:");
+    backtrace().skip(2).for_each(|frame| {
```
Comment (abrestic-rivos) [inline src/main.rs:93]: put a comment about the `skip(2)`

File: src/backtrace.rs (hunk 21 above)
Comment (atulkharerivos) [reply src/backtrace.rs]: Think that as originally suggested, the condition protected against the subtract underflow in ` (fp - stack_start - 2)`. However, with this change, we do need to use `checked_sub()`. One way to do it might be to use the following instead, which will set offset to 0 if we underflow:

` let offset = ((fp.checked_sub(self.stack_start).unwrap_or(2) - 2) / (size_of::<u64>() as u64)) as usize;`

```
    if fp < (stack_start+2) || fp > stack_end {
        break;
    }
```

File: src/backtrace.rs (hunk 25 above)
Comment (atulkharerivos) [reply src/backtrace.rs:77]: Yeah, seems like we will encounter this because we removed the check for ` fp - 2` that guarded against the underflow. Using something like the following should set offset to 0 without a panic, but we may want to check for the underflow and simply bail.

` let offset = ((fp.checked_sub(self.stack_start).unwrap_or(2) - 2) / (size_of::<u64>() as u64)) as usize;`

File: src/backtrace.rs (hunk 23 above)
Comment (stillson) [reply src/backtrace.rs]: My bad. You are right, although I had to add a few *'s

File: src/main.rs (hunk 26 above)
Comment (stillson) [reply src/main.rs:94]: fixed

File: src/backtrace.rs [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+        Self {
+            fp: Some(fp),
+            stack_start,
+            stack_end,
+            stack,
+        }
+    }
+
+    fn next_address(&self, fp: u64) -> Option<(u64, u64)> {
+        let offset = (fp - self.stack_start - 2) as usize / size_of::<u64>();
```
Comment (atulkharerivos) [inline src/backtrace.rs:67]: Needs to guard against potential underflows:

`let offset = fp.checked_sub(self.stack_start).unwrap_or(2) .checked_sub(2).unwrap_or(0) as usize / sizeof::u64();`

File: src/backtrace.rs [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+        if offset == 0 {
```
Comment (atulkharerivos) [inline src/backtrace.rs:68]: Nit: Can elide return with: `if offset > 0 {}...else { None}`

File: src/backtrace.rs (hunk 27 above)
Comment (stillson) [reply src/backtrace.rs:67]: It uses slice.get(), which returns None on out of bounds.

Comment (atulkharerivos) [reply src/backtrace.rs:67]: Think we are referring to different portions of code unless the code needs to be refreshed. In the below, it seems offset is computed unconditionally, which can cause an underflow if `fp == self.stack_start` for example. The first checked_sub() guards against an underflow with `(fp - self.stack_start)`, and the second guards against the case where `'fp - self.stack_start'` evaluates to 0 (which will cause a underflow with the -2).

In either of those cases, offset will be 0.

   ```suggestion
         fn next_address(&self, fp: u64) -> Option<(u64, u64)> {
             let offset = (fp.checked_sub(self.stack_start).unwrap_or(2).checked_sub(2).unwrap_or(0) as usize / size_of::<u64>();
```

File: src/backtrace.rs [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+            stack_end,
+            stack,
+        }
+    }
+
+    fn next_address(&self, fp: u64) -> Option<(u64, u64)> {
+        // if offset is out of bounds get will return none, and the whole function
+        // will return none
+        let offset = ((fp - self.stack_start) / size_of::<u64>() as u64) as usize;
+        let address = *self.stack.get(offset - 1)?;
```
Comment (atulkharerivos) [inline src/backtrace.rs:70]: The issue here is the unchecked subtract.

To illustrate the problem, here's a sample program that demonstrates what happens if offset is 0 for example (it's silly to prevent the compiler from detecting the attempted subtract from 0 at compile time):

```
fn get_offset(b : bool) -> usize {
    if b {
        0
     } else { 4 }
}

fn main() {
    let stack = &[0, 1, 2, 3];
    let address = stack.get(foo(true) - 2);
    println!("{:?}", address);
}
```

File: src/backtrace.rs [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+    fp: Option<u64>,
+    stack_start: u64,
+    stack_end: u64,
+    stack: &'static [u64],
+}
+
+impl BackTrace {
+    fn new(fp: u64) -> Self {
+        let stack_start = HYP_STACK_BOTTOM;
+        let stack_end = HYP_STACK_TOP;
```
Comment (glg-rv) [inline src/backtrace.rs:46]: Now that the stack address is fixed, the stack_start and stack_end could _possibly_ be removed. The slice would be fixed, everything would be more static.

File: src/backtrace.rs [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+        let stack = unsafe {
+            core::slice::from_raw_parts(
+                stack_start as *const u64,
+                (stack_end - stack_start) as usize,
+            )
+        };
+
+        Self {
+            fp: Some(fp),
```
Comment (glg-rv) [inline src/backtrace.rs:55]: I think the best thing to do is to check whether fp is between HYP_STACK_TOP and HYP_STACK_BOTTOM. If not, disable the backtrace.

This will both fix the problem that is unaddressed right now, the fact that before we switch to the VA stack (or when we are on the overflow stack) this would fail, and it would also make the calculation of the offset below safe.

**Review by glg-rv — CHANGES_REQUESTED** [review]
As discussed, this still doesn't handle the problem that stack during the early boot (when we could panic).

File: src/backtrace.rs (hunk 29 above)
Comment (stillson) [reply src/backtrace.rs:66]: Since this is really just 2 addresses (stack_start and stack_start + 8) I guaranteed alignment and check for the addresses. I think that is enough of a check to fix the problem.

File: src/backtrace.rs (hunk 30 above)
Comment (stillson) [reply src/backtrace.rs:46]: we need stack start for bounds checking and calculating the offset, but we don't need stack end.

Comment (stillson) [reply src/backtrace.rs:46]: actually, we use stack_end too. we don't really need it, but I'd rather fail early.

File: src/backtrace.rs (hunk 31 above)
Comment (stillson) [reply src/backtrace.rs:55]: I put a check in, and if it's out of bounds I just return a "NULL" object.

**Review by atulkharerivos — COMMENTED** [review]
Looks good, with the new check for  `if fp <= self.stack_start + size_of::<u64>() as u64 `. BTW, since the stack start/end are constants, can we add a test to check for a boundary conditions? Basically, something like constructing BackTrace with ` fp == HYP_STACK_BOTTOM` and perhaps, 'fp > HYP_STACK_TOP', and verifying that  `.next()` returns `None`.

**Review by atulkharerivos — APPROVED** [review]
Looks like we can't use the actual Salus stack constant values for BT testing, so a test is moot unless we can run it under QEMU.

**Comment by stillson** [comment]
For what it's worth, I started making this testable in another branch, and it work fine.

File: src/backtrace.rs [hunk 32]
```diff
@@ -0,0 +1,115 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::hyp_layout::{HYP_STACK_BOTTOM, HYP_STACK_TOP};
```
Comment (abrestic-rivos) [inline src/backtrace.rs:5]: nit: this `use` should go below with the others.

File: src/main.rs (hunk 16 above)
Comment (abrestic-rivos) [reply src/main.rs:84]: not fixed?

**Review by abrestic-rivos — COMMENTED** [review]
maybe i'm missing something, but it still doesn't look like this is addressing @glg-rv's point about panics that occur in early boot, before the stack is mapped at `HYP_STACK_BOTTOM`.

File: src/main.rs (hunk 26 above)
Comment (stillson) [reply src/main.rs:88]: My bad. I forgot to push out the latest version.

File: src/backtrace.rs [hunk 33]
```diff
@@ -0,0 +1,115 @@
...
+    stack_start: u64,
+    stack_end: u64,
+    stack: &'static [u64],
+}
+
+impl BackTrace {
+    fn new(fp: u64) -> Self {
+        let stack_start = HYP_STACK_BOTTOM;
+        let stack_end = HYP_STACK_TOP;
+        let stack = unsafe {
```
[the remark this answers is not in the document]
Comment (stillson) [reply src/backtrace.rs:47]: Somehow, I had done this and lost it again. Fixed.

**Review by glg-rv — APPROVED** [review]
> maybe i'm missing something, but it still doesn't look like this is addressing @glg-rv's point about panics that occur in early boot, before the stack is mapped at `HYP_STACK_BOTTOM`.

This has since been fixed. Essentially now we check that we are in the HYP_STACK range and if so we attempt the stack backtrace. It is a bit flimsy but it's really hard to know from an exception handler whether we're still in the init phase. (Can't trust PerCpu, etc).

Something we could do in the future is to actually have two different rust exception handler, one for the init (early?) phase and the other from when we switched to hyp paging. We already have two different assembler entry points for the phase so that should be okay.

The reason I prefer not to have a stack trace in the init stage is because it's really hard to know what cpu we're running (could use SBI, but then we have to map it to CpuId and find what stack range we're using) and we really have no protection on the stack, so could also be not trusted.

Regarding this patch, A few things might be done simpler (no reason to keep stack_start and stack_end as structure of backtrace), and also should return none in case stack is not in range. But these are minor details and possibly be left there for trying to support the `init` phase (which I think we don't need).
